// @mui/x-charts/LineChart/seriesConfig/curveEvaluation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.evaluateCurveY = evaluateCurveY;
var _getCurve = require("../../internals/getCurve");
var _cubiqSolver = require("../../internals/cubiqSolver");
/**
 * A straight line segment.
 */

/**
 * A cubic bezier segment with control points.
 */

function isBezierSegment(segment) {
  return 'cpx1' in segment;
}

/**
 * A minimal d3 path context that captures line/bezier segments
 * instead of producing an SVG path string.
 */
class SegmentCapture {
  segments = [];
  cx = 0;
  cy = 0;
  moveTo(x, y) {
    this.cx = x;
    this.cy = y;
  }
  lineTo(x, y) {
    this.segments.push({
      x0: this.cx,
      y0: this.cy,
      x1: x,
      y1: y
    });
    this.cx = x;
    this.cy = y;
  }
  bezierCurveTo(cpx1, cpy1, cpx2, cpy2, x, y) {
    this.segments.push({
      x0: this.cx,
      y0: this.cy,
      cpx1,
      cpy1,
      cpx2,
      cpy2,
      x1: x,
      y1: y
    });
    this.cx = x;
    this.cy = y;
  }
  closePath() {}
}

/** Evaluate a cubic Bezier at parameter t. */
function cubicBezier(t, p0, p1, p2, p3) {
  const mt = 1 - t;
  return mt * mt * mt * p0 + 3 * mt * mt * t * p1 + 3 * mt * t * t * p2 + t * t * t * p3;
}

/**
 * Get polynomials coefficient of a cubic Bezier curve.
 * P(t) = rep[0] * t**3  + rep[1] * t**2 + rep[2] * t + rep[3]
 */
function cubicBezierCoeffs(p0, p1, p2, p3) {
  return [-p0 + 3 * p1 - 3 * p2 + p3, 3 * p0 - 6 * p1 + 3 * p2, -3 * p0 + 3 * p1, p0];
}

/**
 * Find parameter t such that the segment's x(t) ≈ targetX
 */
function findTForX(segment, targetX) {
  if (!isBezierSegment(segment)) {
    // Linear segment.
    const dx = segment.x1 - segment.x0;
    return dx === 0 ? 0 : (targetX - segment.x0) / dx;
  }
  const xBezierCoeffs = cubicBezierCoeffs(segment.x0, segment.cpx1, segment.cpx2, segment.x1);
  const polyToSolve = [...xBezierCoeffs];
  polyToSolve[3] -= targetX;
  const roots = (0, _cubiqSolver.cubicRoots)(polyToSolve);
  if (roots.length > 0) {
    return roots[0];
  }
  return -1;
}

/** Evaluate the segment's y at parameter t. */
function evaluateSegmentY(segment, t) {
  if (!isBezierSegment(segment)) {
    return segment.y0 + t * (segment.y1 - segment.y0);
  }
  return cubicBezier(t, segment.y0, segment.cpy1, segment.cpy2, segment.y1);
}

/**
 * Build the curve segments for a set of pixel-coordinate points
 * using d3's curve factory, then evaluate y at the given pixel x.
 *
 * Returns null if targetX is outside the curve's x range.
 */
function evaluateCurveY(points, targetX, curveType) {
  if (points.length === 0) {
    return null;
  }
  if (points.length === 1) {
    return points[0].y;
  }
  const capture = new SegmentCapture();
  const factory = (0, _getCurve.getCurveFactory)(curveType);
  const curveInstance = factory(capture);
  curveInstance.lineStart();
  for (const p of points) {
    curveInstance.point(p.x, p.y);
  }
  curveInstance.lineEnd();

  // Find the segment containing targetX.
  for (const segment of capture.segments) {
    if (targetX < segment.x0 + 0.5 && targetX > segment.x0 - 0.5) {
      return segment.y0;
    }
    if (targetX < segment.x1 + 0.5 && targetX > segment.x1 - 0.5) {
      return segment.y1;
    }
    const xMin = Math.min(segment.x0, segment.x1);
    const xMax = Math.max(segment.x0, segment.x1);
    if (targetX >= xMin && targetX <= xMax) {
      const t = findTForX(segment, targetX);
      return evaluateSegmentY(segment, t);
    }
  }
  return null;
}
